test/unit/ui/map.test.js

```diff
@@ -523,6 +523,69 @@ test('Map', (t) => {
             t.end();
         });
 
+        t.test('should apply different styles when toggling setStyle (https://github.com/mapbox/mapbox-gl-js/issues/11939)', (t) => {
+            const styleWithTerrainExaggeration = {
+                'terrain': {
+                    'source': 'mapbox-dem',
+                    'exaggeration': 500
+                }
+            };
+
+            const styleWithoutTerrainExaggeration = {
+                'terrain': {
+                    'source': 'mapbox-dem'
+                }
+            };
+
+            const map = createMap(t, {style: extend(createStyle(), styleWithTerrainExaggeration)});
+
+            map.on('load', () => {
+                t.equal(map.style.terrain.properties._values.exaggeration, 500);
+
+                map.setStyle(styleWithoutTerrainExaggeration);
+                t.equal(map.style.terrain.properties._values.exaggeration, 1);
+
+                map.setStyle(styleWithTerrainExaggeration);
+                t.equal(map.style.terrain.properties._values.exaggeration, 500);
+
+                t.equal(styleWithoutTerrainExaggeration.sources.terrain.exaggeration, undefined);
+                t.equal(styleWithTerrainExaggeration.sources.terrain.exaggeration, 500);
+                t.end();
+            });
+
+            map.remove();
+            t.end();
+        });
+
+        t.test('should apply different projections when toggling setStyle (https://github.com/mapbox/mapbox-gl-js/issues/11916)', (t) => {
+            const styleWithWinkelTripel = {
+                'projection': {'name': 'winkelTripel'}
+            };
+
+            const styleWithGlobe = {
+                'projection': {'name': 'globe'}
+            };
+
+            const map = createMap(t, {style: extend(createStyle(), styleWithWinkelTripel)});
```

Comment:
This is creating a new style here, which is not actually testing the bug.
```suggestion
            const map = createMap(t, {style: styleWithWinkelTripel});
```